func/code.func

```diff
@@ -92,73 +95,90 @@ _ unpack_order(slice order) inline_ref {
-    ;; lottery inactive if balance < 2 stake + 1 TON
-    if (my_tons() < o_amount * 2 + 1000000000) {
-       send_tons(player, msg_value, 2);
-    } else {
-      (int o_seqno, int n_wins, int in_amount, int out_amount, int hash) = unpack_stats(stats);
-      ;; Calculating new hash
-      hash = cell_hash(begin_cell()
-        .store_uint(hash, 256)
-        .store_uint(msg_hash, 256)
-        .store_uint(cur_lt(), 64)
-        .store_uint(now(), 64)
-        .end_cell()
-      );
-      randomize(hash);
-      if (rand(10) >= 4) {
-        builder order = pack_order(o_seqno, 3, now(), o_amount, player);
-        orders~udict_set_builder(64, o_seqno, order);
-        in_amount += o_amount;
-        if (o_amount > 3000000000) {
-            send_tons(owner1, o_amount / 6, 3);
-            send_tons(owner2, o_amount / 6, 3);
-        }
-      } else {
-        builder order = pack_order(o_seqno, 2, now(), o_amount, player);
-        orders~udict_set_builder(64, o_seqno, order);
-
-        in_amount += o_amount;
-        out_amount += o_amount * 2;
-        n_wins += 1;
-
-        send_tons(player, 2 * o_amount, 3);
-      }
-      orders = remove_old_orders(orders, o_seqno);
-      o_seqno += 1;
+  slice msg_body = in_msg_body;
+  int op = in_msg_body~load_uint(32);
 
-      set_data(pack_state(o_seqno, n_wins, in_amount, out_amount, hash, owner1, owner2, echo, orders));
-    }
-    return();
+  ;; new random number received from vrf
+  if (op == 0x069CECA8 & slices_equal?(sender, ecvrf)) {
+    int vrfrand = msg_body~load_uint(256);
+    rand_no = vrfrand % 10;
+    rand_updated = -1;
+    return ();
   }
 
-  ;; workchain protection
-  throw_unless(111, workchains_equal?(sender, my_address()));
-
   ;; min stake 1 TON
   throw_unless(333, msg_value > 999999999);
 
   ;; lottery active if balance > 2 stake + 1 TON
   throw_unless(666, my_tons() > msg_value * 2 + 1000000000);
 
-  ;; send order amount to echo contract (crossworkchain)
-  builder msg = begin_cell()
-    .store_uint(0x18, 6)
-    .store_slice(echo)
-    .store_coins(0)
-    .store_uint(0, 1 + 4 + 4 + 64 + 32 + 1 + 1)
-    .store_slice(sender)
-    .store_coins(msg_value);
-  send_raw_message(msg.end_cell(), 64);
+  ;; fallback incase random number is not updated by vrf
+  if (rand_updated == 0) {
+    rand_no = rand(10);
+  }
+
+  (slice player, int o_amount) = (in_msg_body~load_msg_addr(), in_msg_body~load_coins());
+  ;; lottery inactive if balance < 2 stake + 1 TON
+  if (my_tons() < o_amount * 2 + 1000000000) {
+      send_tons(player, msg_value, 2);
+  } else {
+    (int o_seqno, int n_wins, int in_amount, int out_amount, int hash) = unpack_stats(stats);
+    ;; Calculating new hash
+    hash = cell_hash(begin_cell()
+      .store_uint(hash, 256)
+      .store_uint(msg_hash, 256)
+      .store_uint(cur_lt(), 64)
+      .store_uint(now(), 64)
+      .end_cell()
+    );
+    randomize(hash);
+    if (rand_no >= 4) {
```

Comment:
There is an issue since player can know random number stored in the contract and thus knows whether the bet is winning or losing.

Reply:
Actually I was thinking the get_random function as a way of testing if random numbers are actually being updated. Could there be any other way of testing?